src/cert/chain.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Cert chains.
+
+use core::num::NonZeroUsize;
+
+use arrayvec::ArrayVec;
+
+use crate::cert::Cert;
+use crate::cert::CertFormat;
+use crate::cert::Error;
+use crate::crypto::sig;
+use crate::io;
+
+/// A trust chain collection.
+///
+/// A trust chain consists of a sequence of certificates, starting with a
+/// self-signed root certificate, and ending in a leaf certificate with some
+/// a public key we wish to authenticate against the root.
+///
+/// A `TrustChain` actually managed a small number of chains across several
+/// *slots*; all Cerberus authentication messages refer to a specific
+/// certificate slot.
+pub trait TrustChain {
+    /// Gets the length of the `slot`th chain.
+    ///
+    /// Cannot be zero; returns `None` if this chain has no such slot.
+    fn chain_len(&self, slot: u8) -> Option<NonZeroUsize>;
+
+    /// Gets the `index`th cert of the `slot`th chain.
+    ///
+    /// Returns `None` if `index` is out of bounds or if there is no `slot`th
+    /// chain. These cases can be distinguished by calling `chain_len()`.
+    fn cert(&self, slot: u8, index: usize) -> Option<&Cert>;
+}
+
+/// A simple trust with only one slot.
+#[derive(Debug)]
+pub struct SimpleChain<'cert, const LEN: usize> {
+    chain: [Cert<'cert>; LEN],
+}
+
+impl<'cert, const LEN: usize> SimpleChain<'cert, LEN> {
+    /// Parses and verifies the trust chain described by `raw_chain`,
+    /// which starts at the root certificate for the trust chain.
+    pub fn parse(
+        raw_chain: &[&'cert [u8]],
+        format: CertFormat,
+        ciphers: &mut impl sig::Ciphers,
+    ) -> Result<Self, Error> {
+        if raw_chain.len() != LEN {
+            return Err(Error::Io(io::Error::BufferExhausted));
```

Comment:
Returning `BufferExhausted` if a `raw_chain` that is shorter than `LEN` is passed seems misleading. Is this error code from the Cerberus specs, or is it something we can change?

Reply:
Caught me being lazy, hehe.